File: scenes/ClockScene.ts
import { Font } from 'rpi-led-matrix';
import Scene from "./Scene";

class ClockScene extends Scene {
  start(matrix) {
    const timeString = (new Date()).toTimeString();
    const [hoursString, minutesString] = timeString.split(':');
    const hoursInt = parseInt(hoursString);
    const minutesInt = parseInt(minutesString);
    const minutesRounded = Math.round(minutesInt / 5) * 5;
    const hours = (hoursInt + (minutesRounded >= 60 ? 1 : 0)) % 24;
    const minutes = minutesRounded % 60;

    const hoursWords = {
      0: 'MIDNIGHT',
      1: 'ONE',
      2: 'TWO',
      3: 'THREE',
      4: 'FOUR',
      5: 'FIVE',
      6: 'SIX',
      7: 'SEVEN',
      8: 'EIGHT',
      9: 'NINE',
      10: 'TEN',
      11: 'ELEVEN',
      12: 'NOON',
      13: 'ONE',
      14: 'TWO',
      15: 'THREE',
      16: 'FOUR',
      17: 'FIVE',
      18: 'SIX',
      19: 'SEVEN',
      20: 'EIGHT',
      21: 'NINE',
      22: 'TEN',
      23: 'ELEVEN',
      24: 'MIDNIGHT',
    };

    const minutesWords = {
      5: 'FIVE PAST',
      10: 'TEN PAST',
      15: 'QUARTER PAST',
      20: 'TWENTY PAST',
      25: 'TWENTY-FIVE PAST',
      30: 'HALF PAST',
      35: 'TWENTY-FIVE TIL',
      40: 'TWENTY TIL',
      45: 'QUARTER TIL',
      50: 'TEN TIL',
      55: 'FIVE TIL',
    }

    let one = '';
    let two = hoursWords[hours];
    let three = '';
    
    if (minutes) {
      [one, two] = minutesWords[minutes].split(' ');
      if (minutes <= 30) {
        three = hoursWords[hours]
      } else {
        three = hoursWords[hours + 1];
      }
    }

    console.log([one, two, three].join(' '));

    const font = new Font('4x6', `${process.cwd()}/node_modules/rpi-led-matrix/fonts/4x6.bdf`);
    matrix.font(font).fgColor(0xffffff);
    matrix.drawText(one, 5, 5);
    matrix.drawText(two, 10, 14);
    matrix.drawText(three, 15, 23);
  }
}

File: src/main/java/dev/jchristoffersen/lightbox/scenes/ClockScene.java
package dev.jchristoffersen.lightbox.scenes;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

import dev.jchristoffersen.lightbox.constants.Constants;
import dev.jchristoffersen.lightbox.render.FrameBuffer;
import dev.jchristoffersen.lightbox.scene.StaticScene;
import dev.jchristoffersen.lightbox.text.ParsedBdf;
import dev.jchristoffersen.lightbox.text.TextRenderer;

public class ClockScene extends StaticScene {
    private static final Map<Integer, String> hoursWords = Map.ofEntries(
        Map.entry(0, "MIDNIGHT"),
        Map.entry(1, "ONE"),
        Map.entry(2, "TWO"),
        Map.entry(3, "THREE"),
        Map.entry(4, "FOUR"),
        Map.entry(5, "FIVE"),
        Map.entry(6, "SIX"),
        Map.entry(7, "SEVEN"),
        Map.entry(8, "EIGHT"),
        Map.entry(9, "NINE"),
        Map.entry(10, "TEN"),
        Map.entry(11, "ELEVEN"),
        Map.entry(12, "NOON"),
        Map.entry(13, "ONE"),
        Map.entry(14, "TWO"),
        Map.entry(15, "THREE"),
        Map.entry(16, "FOUR"),
        Map.entry(17, "FIVE"),
        Map.entry(18, "SIX"),
        Map.entry(19, "SEVEN"),
        Map.entry(20, "EIGHT"),
        Map.entry(21, "NINE"),
        Map.entry(22, "TEN"),
        Map.entry(23, "ELEVEN")
    );

    private static final Map<Integer, String> minutesWords = Map.ofEntries(
        Map.entry(5, "FIVE PAST"),
        Map.entry(10, "TEN PAST"),
        Map.entry(15, "QUARTER PAST"),
        Map.entry(20, "TWENTY PAST"),
        Map.entry(25, "TWENTY-FIVE PAST"),
        Map.entry(30, "HALF PAST"),
        Map.entry(35, "TWENTY-FIVE TILL"),
        Map.entry(40, "TWENTY TILL"),
        Map.entry(45, "QUARTER TILL"),
        Map.entry(50, "TEN TILL"),
        Map.entry(55, "FIVE TILL")
    );

    private ParsedBdf parsedBdf;

    public CompletableFuture<Void> prep() {
        return CompletableFuture.runAsync(() -> {
            this.parsedBdf = ParsedBdf.parse("TODO");
        });
    }

    public FrameBuffer renderOnce() {
        FrameBuffer frameBuffer = new FrameBuffer(Constants.WIDTH, Constants.HEIGHT);

        int hours = LocalDateTime.now().getHour();
        int minutes = LocalDateTime.now().getMinute();
        int minutesRounded = ((int) Math.round((double) minutes / 5) * 5) % 60;

        // render up to three words, each in their own row
        // by default render the hour on the middle line
        String one = "";
        String two = hoursWords.get(hours);
        String three = "";

        // if there are minutes, render the minutes words as well
        if (minutesRounded > 0) {
            String[] minutesWordsArray = minutesWords.get(minutesRounded).split("\s");
            one = minutesWordsArray[0];
            two = minutesWordsArray[1];
            if (minutesRounded <= 30) {
                three = hoursWords.get(hours);
            } else {
                three = hoursWords.get(hours + 1);
            }
        }

        TextRenderer textRenderer = new TextRenderer(parsedBdf, 0xffffff);
        textRenderer.renderTextToFrameBuffer(one, 0, -1, frameBuffer);
        textRenderer.renderTextToFrameBuffer(two, 0, 7, frameBuffer);
        textRenderer.renderTextToFrameBuffer(three, 0, 14, frameBuffer);

        return frameBuffer;
    }
}
